<a href="https://colab.research.google.com/github/Jenesha20/GenAI_Self_Learn/blob/main/rag.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install -q langchain langchain-community langchain-huggingface

In [ ]:
!pip install -q chromadb pypdf sentence-transformers

In [ ]:
!pip install -q transformers accelerate torch

In [ ]:
from langchain_community.document_loaders import PyPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma

In [ ]:
from google.colab import files

uploaded = files.upload()
pdf_path = list(uploaded.keys())[0]


Saving Jenesha_Malar.pdf to Jenesha_Malar.pdf


In [ ]:
loader = PyPDFLoader(pdf_path)
documents = loader.load()

print(f"Total pages loaded: {len(documents)}")
print(documents[0].metadata)


Total pages loaded: 1
{'producer': 'xdvipdfmx (20240305)', 'creator': 'LaTeX with hyperref', 'creationdate': '2025-04-18T02:57:56+00:00', 'source': 'Jenesha_Malar.pdf', 'total_pages': 1, 'page': 0, 'page_label': '1'}


In [ ]:
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100
)

chunks = text_splitter.split_documents(documents)
print(f"Total chunks created: {len(chunks)}")
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}:\n{chunk.page_content}\n---\n")

Total chunks created: 7
Chunk 1:
Jenesha Malar
jeneshamalars@gmail.com | +91 8870770296
EDUCATION
BTECH AI AND DS
KONGU ENGINEERING
COLLEGE
Grad. May 2026 | CGPA : 9.06/ 10
HSC
CJM MATRIC HR.SEC SCHOOL
Grad. May 2022 | Per : 94.33
SSLC
CJM MATRIC HR.SEC SCHOOL
Grad. May 2020 | Per : 84
LINKS
Github:// Jenesha20
LinkedIn:// Jenesha Malar
Leetcode:// JeneshaMalar
AREAS OF INTEREST
Database Management System
Data Science
SKILLS
TECHNICAL
Languages:
• Java • C • SQL • Python
Frameworks & Tools:
• Tableau • Power BI • Keras • Numpy
---

Chunk 2:
Languages:
• Java • C • SQL • Python
Frameworks & Tools:
• Tableau • Power BI • Keras • Numpy
• Pandas • GIT • Figma • Linux
DevOps:
• Docker
ACHIEVEMENTS
• Secured Academic Excellence Award
• 1st in Paper Presentation at Autonix
• 2nd in Paper Presentation at MarkUs
• Secured award for Centum in CS
LEADERSHIP
Treasurer AICC (2024 - 25)
Joint Secretary AICC (2023 - 24)
Executive Member SDC (2023 - 24)
PUBLICATIONS
IEEE Enhancing Disease Symptom
Pred

In [ ]:
for i, chunk in enumerate(chunks):
    chunk.metadata["chunk_id"] = i
    print(chunk.metadata["chunk_id"])

0
1
2
3
4
5
6


In [ ]:
embedding_model = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
persist_directory = "./chroma_db"
vectorstore = Chroma.from_documents(
    documents=chunks,
    embedding=embedding_model,
    persist_directory=persist_directory
)


vectorstore.persist()
print(f"Vector store persisted to: {persist_directory}")

Vector store persisted to: ./chroma_db


/tmp/ipython-input-4108207497.py:11: LangChainDeprecationWarning: Since Chroma 0.4.x the manual persistence method is no longer supported as docs are automatically persisted.
  vectorstore.persist()


In [ ]:
chroma_collection = vectorstore._collection

# Get all IDs from the collection (or a sample of IDs if the collection is very large)
# LangChain's Chroma implementation assigns UUIDs as internal IDs by default.
# To get actual stored IDs, we can query the collection.

retrieved_info = chroma_collection.peek(limit=3)

# Check if embeddings are present by checking the length of the array
if retrieved_info and len(retrieved_info['embeddings']) > 0:
    print(f"Displaying {len(retrieved_info['embeddings'])} sample embeddings from the vector database:")
    for i in range(len(retrieved_info['embeddings'])):
        print(f"\n--- Entry {i+1} ---")
        print(f"ID: {retrieved_info['ids'][i]}")
        print(f"Document (first 200 chars):\n{retrieved_info['documents'][i][:200]}...")
        print(f"Embedding (first 10 values):\n{retrieved_info['embeddings'][i][:10]}...")
        print(f"Embedding length: {len(retrieved_info['embeddings'][i])}")
else:
    print("Could not retrieve sample embeddings. The collection might be empty or an error occurred.")

Displaying 3 sample embeddings from the vector database:

--- Entry 1 ---
ID: 72313a26-c443-4b67-ba72-fb2ae26de7bf
Document (first 200 chars):
Jenesha Malar
jeneshamalars@gmail.com | +91 8870770296
EDUCATION
BTECH AI AND DS
KONGU ENGINEERING
COLLEGE
Grad. May 2026 | CGPA : 9.06/ 10
HSC
CJM MATRIC HR.SEC SCHOOL
Grad. May 2022 | Per : 94.33
SS...
Embedding (first 10 values):
[-0.0291125  -0.03219274  0.00444849  0.0177064  -0.05716434 -0.09255209
  0.01964277  0.01212048 -0.16547939  0.05357023]...
Embedding length: 384

--- Entry 2 ---
ID: ac0004dd-9581-429c-9c21-df6ef287de6e
Document (first 200 chars):
Languages:
• Java • C • SQL • Python
Frameworks & Tools:
• Tableau • Power BI • Keras • Numpy
• Pandas • GIT • Figma • Linux
DevOps:
• Docker
ACHIEVEMENTS
• Secured Academic Excellence Award
• 1st in ...
Embedding (first 10 values):
[-1.90342162e-02  4.94152773e-05  3.75780910e-02  3.09005473e-02
  6.57306015e-02 -3.17945443e-02  3.02234590e-02  7.08924681e-02
 -1.13156766e-01 -2.7977163

In [ ]:
query = "What are the experience?"

results = vectorstore.similarity_search_with_score(
    query,
    k=3
)

for doc, score in results:
    print("Similarity Score:", score)
    print("Metadata:", doc.metadata)
    print("Content:", doc.page_content[:300])
    print("-" * 80)


Similarity Score: 1.5950535535812378
Metadata: {'page_label': '1', 'total_pages': 1, 'creator': 'LaTeX with hyperref', 'chunk_id': 3, 'creationdate': '2025-04-18T02:57:56+00:00', 'page': 0, 'producer': 'xdvipdfmx (20240305)', 'source': 'Jenesha_Malar.pdf'}
Content: EXPERIENCE
PUNCHBIZ | Full Stack Data Scientist
July 2024 – Dec 2024 | Intern
• Built a real-time Intruder Detection System with Flask backend and React
Native for web & mobile.
• Integrated YOLO for person detection and email alerts, enhancing security
monitoring.
INTEL | Full Stack Developer
May 2
--------------------------------------------------------------------------------
Similarity Score: 1.602097749710083
Metadata: {'source': 'Jenesha_Malar.pdf', 'creator': 'LaTeX with hyperref', 'page_label': '1', 'page': 0, 'producer': 'xdvipdfmx (20240305)', 'chunk_id': 4, 'creationdate': '2025-04-18T02:57:56+00:00', 'total_pages': 1}
Content: • Implemented user authentication and CRUD operations to enhance website
functionality.

### Understanding Similarity Scores

In a vector database like Chroma, similarity scores quantify how relevant a retrieved document (or chunk of text) is to a given query. The scores are derived from the similarity between the vector embeddings of the query and the document chunks.

#### How it Works Internally:

1.  **Embedding Generation**: Both your query (`"What are the experience?"`) and each document chunk (from your PDF) are converted into numerical vectors, called embeddings. These embeddings are generated by the `HuggingFaceEmbeddings` model (`"sentence-transformers/all-MiniLM-L6-v2"`). The model is trained to produce embeddings where semantically similar texts are closer to each other in the vector space.

2.  **Vector Space Representation**: Imagine a high-dimensional space where each word, phrase, or document chunk is a point. Embeddings are the coordinates of these points. Texts with similar meanings are mapped to points that are geometrically closer together.

3.  **Similarity Calculation**: When you perform a `similarity_search_with_score` query, the embedding of your query is compared to the embeddings of all document chunks in the vector store. The "similarity score" is typically calculated using a distance metric, most commonly **cosine similarity** or **Euclidean distance**.
    *   **Cosine Similarity**: Measures the cosine of the angle between two vectors. A score closer to 1 indicates high similarity (small angle), and a score closer to -1 indicates high dissimilarity (large angle). In the context of `sentence-transformers` and `Chroma`, the distance often represents a transformed version of cosine similarity (e.g., `1 - cosine_similarity`), where a lower score indicates higher similarity.
    *   **Euclidean Distance**: Measures the straight-line distance between two points in vector space. A smaller Euclidean distance indicates higher similarity.

4.  **Ranking Results**: The vector database computes these scores for all relevant document chunks and then sorts them by similarity (or distance). The `k` most similar chunks are returned, along with their scores.

In the output above, the `Similarity Score` value indicates the distance between the query embedding and the chunk embedding. A *lower* score generally means *higher* similarity for the distance metric used by `Chroma` with the default `HuggingFaceEmbeddings` (which often implies a lower distance value means closer vectors). For example, a score of `1.595` is considered more similar than `1.636` to the query, indicating that the chunk with `1.595` score is more relevant.

### Understanding Word Embeddings and Similarity Scores

#### What are Word Embeddings?

Word embeddings are numerical representations of words or phrases in a high-dimensional vector space. The core idea is that words with similar meanings will have similar numerical representations (i.e., their vectors will be close to each other in the vector space). These embeddings capture semantic relationships and context.

Historically, techniques like Word2Vec and GloVe learned fixed embeddings for each word. More recently, contextual embedding models (like BERT, GPT, and the underlying models for Sentence-Transformers) generate embeddings that vary based on the word's context within a sentence, providing even richer semantic representations.

#### How are Embeddings Created (Mathematical Intuition)?

While the exact math can be complex, especially for modern neural network-based models, here's a simplified intuition:

1.  **Distributional Hypothesis**: The foundation is the idea that words that appear in similar contexts tend to have similar meanings. Models learn by analyzing large text corpora.
2.  **Prediction Tasks**: Many models learn embeddings by performing a prediction task:
    *   **Word2Vec (Skip-gram)**: Given a word, predict its surrounding context words. The weights learned in the neural network's hidden layer become the word embeddings.
    *   **Word2Vec (CBOW)**: Given context words, predict the target word.
    *   **GloVe**: Constructs a global word-word co-occurrence matrix from the corpus and then factorizes it to obtain dense word vectors. It directly optimizes to capture ratios of co-occurrence probabilities.
3.  **Sentence/Chunk Embeddings (like `sentence-transformers`)**: For `HuggingFaceEmbeddings` and `sentence-transformers`, the model doesn't just embed individual words; it creates a single, dense vector representation for an entire sentence, paragraph, or chunk of text. This is often done by passing the text through a pre-trained transformer model (like BERT or RoBERTa) and then pooling (e.g., averaging) the output embeddings of the individual tokens to get a fixed-size vector for the entire input.

The resulting embedding for a piece of text (like your document chunks) is a vector, such as `[-0.029, -0.032, 0.004, ..., 0.053]`, where each number is a dimension in the vector space.

#### Similarity Calculation with Math (Cosine Similarity):

Once texts are converted into vectors, their similarity can be measured using various distance metrics. **Cosine similarity** is one of the most common and effective for text embeddings because it measures the cosine of the angle between two vectors. It focuses on the orientation of the vectors rather than their magnitude.

Given two vectors, $A$ and $B$, their cosine similarity is calculated as:

\[\text{Cosine Similarity}(A, B) = \frac{A \cdot B}{||A|| \cdot ||B||}\]

Where:
*   $A \cdot B$ is the dot product of vectors $A$ and $B$.
*   $||A||$ is the Euclidean norm (magnitude) of vector $A$.
*   $||B||$ is the Euclidean norm (magnitude) of vector $B$.

**Interpretation:**
*   A cosine similarity of 1 means the vectors are identical in direction (maximum similarity).
*   A cosine similarity of 0 means the vectors are orthogonal (no similarity).
*   A cosine similarity of -1 means the vectors are exactly opposite (maximum dissimilarity).

**How `similarity_search_with_score` uses it:**

Many vector databases and embedding models, when reporting a "score" or "distance," might present a transformed version of cosine similarity. For `sentence-transformers` and Chroma, a common practice is to treat a *lower score* as *higher similarity*. This often corresponds to a *distance* metric, which is inversely related to similarity.

For example, if the underlying calculation is cosine similarity (ranging from -1 to 1):
*   A common distance metric derived from cosine similarity is `1 - Cosine Similarity`. In this case, `1 - 1 = 0` (perfect match, 0 distance) and `1 - (-1) = 2` (perfect mismatch, 2 distance). So, smaller values indicate higher similarity.

The exact range and interpretation of the `Similarity Score` can depend on the specific implementation of the embedding model and vector store. However, the general principle holds: the scoring mechanism is designed such that closer (lower) scores indicate more semantic relevance to your query.

In [ ]:
from transformers import pipeline

llm = pipeline(
    "text2text-generation",
    model="google/flan-t5-base",
    max_length=512
)


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/990M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json: 0.00B [00:00, ?B/s]

Device set to use cpu


In [ ]:
query = "List all the project"

retrieved_docs = vectorstore.similarity_search(
    query,
    k=3
)


In [ ]:
context = "\n\n".join(
    [doc.page_content for doc in retrieved_docs]
)


In [ ]:
prompt = f"""
Answer the question using ONLY the context below.
If the answer is not present, say "Not found in the document."

Context:
{context}

Question:
{query}
"""


In [ ]:
response = llm(prompt)

print("Final Answer:\n")
print(response[0]["generated_text"])


Final Answer:

PROJECTS PERSONALIZED CAREER DEVELOPMENT | Prompting | LLM • Built an AI platform with LLaMA-3-8b-8192 for personalized career • Built an AI platform with LLaMA-3-8b-8192 for personalized career roadmaps, skill guidance, and resume optimization. • Github:// Personalized Career Development HOUSING SERVICES | MERN • Developed a MERN-based housing services website with user interaction and data management features. • Github:// Housing Services DOGS BREED IDENTIFIER | Deep Learning | Flask | Docker • Developed a Flask application utilizing the Xception deep learning EXPERIENCE PUNCHBIZ | Full Stack Data Scientist July 2024 – Dec 2024 | Intern • Built a real-time Intruder Detection System with Flask backend and React Native for web & mobile. • Integrated YOLO for person detection and email alerts, enhancing security monitoring. INTEL | Full Stack Developer May 2024 – July 2024 | Intern • Developed a responsive and dynamic website using the MERN stack


In [ ]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnablePassthrough
from langchain_core.output_parsers import StrOutputParser
from langchain_huggingface.llms import HuggingFacePipeline

In [ ]:
llm_langchain = HuggingFacePipeline(pipeline=llm)

In [ ]:
retriever = vectorstore.as_retriever()

In [ ]:
rag_prompt = ChatPromptTemplate.from_template(
    """Answer the question using ONLY the context below.\nIf the answer is not present, say \"Not found in the document.\"\n\nContext:\n{context}\n\nQuestion:\n{question}\nAnswer:"""
)

In [ ]:
rag_chain = (
    { "context": retriever, "question": RunnablePassthrough() }
    | rag_prompt
    | llm_langchain
    | StrOutputParser()
)

In [ ]:
sample_question = "What are the tech stack listed in the document?"
response = rag_chain.invoke(sample_question)

print("Question:", sample_question)
print("RAG Chain Response:", response)

Question: What are the tech stack listed in the document?
RAG Chain Response: Java • C • SQL • PythonnFrameworks & Tools:n• Tableau • Power BI • Keras • Numpy
